src/lseq.rs

```diff
@@ -0,0 +1,269 @@
+            // bounds respectively if they are not found.
+            let prev = match ix.checked_sub(1) {
+                Some(i) => &self.seq.get(i).unwrap().id,
+                None => &lower,
+            };
+            let next = self
+                .seq
+                .get(ix)
+                .map(|Entry { id, .. }| id)
+                .unwrap_or(&upper);
+            let a = self.gen.alloc(&prev, next);
+
+            assert!(prev < &a);
+            assert!(&a < next);
+            a
+        };
+        let op = Op::Insert {
+            id: ix_ident,
+            dot: self.dot.clone(),
+            c,
+        };
+        self.dot.counter += 1;
+        self.apply(op.clone());
+        op
+    }
+
+    /// Perform a local deletion at `ix`.
+    ///
+    /// If `ix` is out of bounds, i.e. `ix > self.len()`, then
+    /// the `Op` is not performed and `None` is returned.
+    pub fn delete_index(&mut self, ix: usize) -> Option<Op<T, A>>
+    where
+        T: Clone,
+    {
+        if ix >= self.seq.len() {
+            return None;
+        }
+
+        let data = self.seq[ix].clone();
+
+        let op = Op::Delete {
+            id: data.id,
+            remote: data.dot,
+            dot: self.dot.clone(),
+        };
+
+        self.dot.counter += 1;
+        self.apply(op.clone());
+
+        Some(op)
+    }
+
+    /// Perform a local deletion at `ix`. If `ix` is out of bounds
+    /// then the last element will be deleted, i.e. `self.len() - 1`.
+    pub fn delete_index_or_last(&mut self, ix: usize) -> Op<T, A>
+    where
+        T: Clone,
+    {
+        match self.delete_index(ix) {
+            None => self
+                .delete_index(self.len() - 1)
+                .expect("delete_index_or_last: 'self.len() - 1'"),
+            Some(op) => op,
+        }
+    }
+
+    /// Get the length of the LSEQ.
+    pub fn len(&self) -> usize {
+        self.seq.len()
+    }
+
+    /// Check if the LSEQ is empty.
+    pub fn is_empty(&self) -> bool {
+        self.seq.is_empty()
+    }
+
+    /// Get the elements represented by the LSEQ.
+    pub fn iter(&self) -> impl Iterator<Item = &T> + '_ {
+        self.seq.iter().map(|Entry { c, .. }| c)
+    }
+
+    /// Access the internal representation of the LSEQ
+    pub fn raw_entries(&self) -> &Vec<Entry<T, A>> {
+        &self.seq
+    }
+}
+
+impl<T, A: Actor> CmRDT for LSeq<T, A> {
+    type Op = Op<T, A>;
+    /// Apply an operation to an LSeq instance.
+    ///
+    /// If the operation is an insert and the identifier is **already** present in the LSEQ instance
+    /// the result is a no-op
+    ///
+    /// If the operation is a delete and the identifier is **not** present in the LSEQ instance the
+    /// result is a no-op
+    fn apply(&mut self, op: Self::Op) {
+        match op {
+            Op::Insert { id, dot, c } => self.insert(id, dot, c),
+            Op::Delete { id, .. } => self.delete(id),
```

Comment:
hmmm, why do we throw away the rest of the information of the `Op::Delete`, if the fields are not needed why store them in the struct? Are we leaning on the Causality Barrier to ensure idempotency over ops?

Reply:
Yea that aditional information is required to implement the `CausalOp` trait of the `VVwE` pr. 

```
impl CausalOp for Op {
    fn happens_after(&self) -> Option<Dot> {
        match self {
            Op::Insert{..} => None,
            Op::Delete{remote ,..} => Some(remote),
        }
    }

    fn dot(&self) -> Dot {
      match self {
         Op::Insert{dot, .. } => dot,
         Op::Delete{dot, .. } => dot,
     }
  }
}
```

he delete can't happen before the operation `remote` happened. It's ugly but I'm not sure how else to handle this case. 

Really we _only_ need the `counter` of the dot since the actor would also be stored in the identifier we're deleting (since it was the one to insert it). Having access to the counter is important because it's what allows the causality barrier to know which messages have been missed. If we see a `Delete {id = X, remote = (B, 10), dot = (A, 5) }` we know that we have (potentially) missed `(B, 1)... (B, 9)` and `(A, 1)...(A, 4)` which is not information we can learn from the identifier.

Reply:
If the responsibility split between lseq and causality barrier forms a clear data boundary, we can follow a layered model where we wrap lseq op's with a causal op, CausalOp(LSeq::Op). The CausalityBarrier would then unwrap these ops for us to release to the protected LSeq.

I'm not sure the boundary is as clear as we'd like though. I wouldn't want to duplicate causality tracking data across the causal layer and the lseq layer just to have these layered decomposition :/

Reply:
Also I wonder how much safety we want to add to the LSeq itself, we can add idempotoncy quite simply by putting in a top level vclock and checking the op dot against the vclock to see if we've already applied it. (right now an insert followed by a delete followed by a replay of the original insert will end up with the element remaining in the LSeq).